h5pt.go

```diff
@@ -81,40 +93,170 @@ func (t *Table) ReadPackets(start, nrecords int, data interface{}) error {
 	return h5err(err)
 }
 
+func extractStructValues(rv reflect.Value, rt reflect.Type) (ptr unsafe.Pointer, err error) {
+	// Initially the memory will be allocated to 64 bytes. If more is required
+	// later, this will be increased.
+	memSize := C.size_t(64)
+	ptr = C.malloc(memSize)
+
+	for i := 0; i < rv.NumField(); i++ {
+		var dataPtr unsafe.Pointer
+		var dataSize C.size_t
+		f := rv.Field(i)
+		ft := string(f.Type().Name())
+		offset := C.size_t(rt.Field(i).Offset)
+
+		switch ft {
+		case "int8":
+			val := C.int8_t(int8(f.Int()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "uint8":
+			val := C.uint8_t(uint8(f.Uint()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "int16":
+			val := C.uint8_t(int16(f.Int()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "uint16":
+			val := C.uint8_t(uint16(f.Uint()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "int32":
+			val := C.uint8_t(int32(f.Int()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "uint32":
+			val := C.uint8_t(uint32(f.Uint()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "int64":
+			val := C.int64_t(int64(f.Int()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "uint64":
+			val := C.uint64_t(uint64(f.Uint()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "float32":
+			val := C.float(float32(f.Float()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "float64":
+			val := C.double(float64(f.Float()))
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		case "bool":
+			val := C.uchar(0)
+			if f.Bool() {
+				val = 1
+			}
+			dataPtr = unsafe.Pointer(&val)
+			dataSize = C.size_t(unsafe.Sizeof(val))
+		default:
+			err = fmt.Errorf("hdf5: Could not append struct member %s "+
+				"to PacketTable", ft)
+			return nil, err
+		}
+
+		// If the earlier allocated memory is not enough, we will increase it
+		// by another 64 bytes.
+		if memSize < (dataSize + offset) {
+			memSize += 64
+			ptr = C.realloc(ptr, memSize)
+			C.memset(unsafe.Pointer(uintptr(ptr)+uintptr(offset)), 0, 64)
+		}
+		C.memcpy(unsafe.Pointer(uintptr(ptr)+uintptr(offset)), dataPtr, dataSize)
+	}
+
+	return ptr, nil
+}
+
 // Append appends packets to the end of a packet table.
-func (t *Table) Append(data interface{}) error {
+func (t *Table) Append(data interface{}) (err error) {
 	rv := reflect.Indirect(reflect.ValueOf(data))
+	rp := reflect.Indirect(reflect.ValueOf(&data))
 	rt := rv.Type()
-	c_nrecords := C.size_t(0)
-	c_data := unsafe.Pointer(nil)
+	cNrecords := C.size_t(1)
+	cData := unsafe.Pointer(nil)
 
 	switch rt.Kind() {
+	case reflect.Slice, reflect.Array:
```

Comment:
thinking a bit more about this, I realize these new semantics break what (I think) the concept of a packet means for hdf5.

it seems to me it is a valid hdf5 thing to say you want a packet to be 1 slice or 1 array of N float64s.
doing it like we do it now, we unravel each packet of 1 array of N float64s into N packets of 1 float64.

I believe the handling of arrays and slices can be left at it was before.

Reply:
I would've love to keep your code. But it does not pass the test:

> --- FAIL: TestPTStruct (0.00s)
> panic: reflect.Value.UnsafeAddr of unaddressable value [recovered]
>         panic: reflect.Value.UnsafeAddr of unaddressable value
> 
> goroutine 12 [running]:
> testing.tRunner.func1(0xc4200be780)
>         /snap/go/1473/src/testing/testing.go:742 +0x29d
> panic(0x52b920, 0x5797a0)
>         /snap/go/1473/src/runtime/panic.go:505 +0x229
> reflect.Value.UnsafeAddr(0x5281e0, 0xc42000c180, 0x97, 0x7)
>         /snap/go/1473/src/reflect/value.go:1772 +0xc4
> hdf5_write/vendor/gonum.org/v1/hdf5.(*Table).Append(0xc420026800, 0x5281e0, 0xc42000c180, 0x553740, 0xc42001a200)
>         /home/jonas/Repos/rapiddecode/write/src/hdf5_write/vendor/gonum.org/v1/hdf5/h5pt.go:199 +0x709
> hdf5_write/vendor/gonum.org/v1/hdf5.testTable(0xc4200be780, 0x553740, 0xc42001a200, 0x5281e0, 0xc42000c180, 0x7)
>         /home/jonas/Repos/rapiddecode/write/src/hdf5_write/vendor/gonum.org/v1/hdf5/h5pt_test.go:61 +0x3d9
> hdf5_write/vendor/gonum.org/v1/hdf5.TestPTStruct(0xc4200be780)
>         /home/jonas/Repos/rapiddecode/write/src/hdf5_write/vendor/gonum.org/v1/hdf5/h5pt_test.go:127 +0x200
> testing.tRunner(0xc4200be780, 0x5689d8)
>         /snap/go/1473/src/testing/testing.go:777 +0xd0
> created by testing.(*T).Run
>         /snap/go/1473/src/testing/testing.go:824 +0x2e0
> FAIL    hdf5_write/vendor/gonum.org/v1/hdf5     0.008s